File: utils/validators.py
"""Validation utilities."""

import re
from decimal import Decimal, InvalidOperation
from typing import Optional


def validate_amount(amount_str: str) -> Optional[Decimal]:
    """
    Validate and parse amount string.
    
    Args:
        amount_str: Amount string (e.g., "$50", "50.99", "fifty")
        
    Returns:
        Decimal amount or None if invalid
    """
    if not amount_str:
        return None
    
    # Remove currency symbols and whitespace
    cleaned = re.sub(r'[\$€£¥,\s]', '', amount_str.strip())
    
    try:
        amount = Decimal(cleaned)
        if amount <= 0:
            return None
        return amount
    except (InvalidOperation, ValueError):
        return None


def validate_phone_number(phone: str) -> bool:
    """
    Validate phone number format.
    
    Args:
        phone: Phone number string
        
    Returns:
        True if valid, False otherwise
    """
    if not phone:
        return False
    
    # Remove common formatting characters
    cleaned = re.sub(r'[\s\-\(\)\+]', '', phone)
    
    # Basic validation: should contain only digits and optional + prefix
    pattern = r'^\+?\d{10,15}$'
    return bool(re.match(pattern, cleaned))


def validate_category_name(name: str) -> tuple[bool, Optional[str]]:
    """
    Validate category name.
    
    Args:
        name: Category name
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not name:
        return False, "Category name cannot be empty"
    
    name = name.strip()
    
    if len(name) < 1:
        return False, "Category name cannot be empty"
    
    if len(name) > 100:
        return False, "Category name cannot exceed 100 characters"
    
    # Check for invalid characters
    if re.search(r'[<>:"/\\|?*]', name):
        return False, "Category name contains invalid characters"
    
    return True, None

File: src/utils/validators.js
/**
 * Validation utilities
 */

/**
 * Validate and parse amount string
 * @param {string} amountStr - Amount string
 * @returns {number|null} Parsed amount or null if invalid
 */
function validateAmount(amountStr) {
  if (!amountStr) return null;
  
  // Remove currency symbols and whitespace
  let cleaned = amountStr.toString().trim();
  cleaned = cleaned.replace(/[$€£¥Rp\s,]/gi, '');
  
  // Handle 'k' suffix (e.g., 50k = 50000)
  if (/^\d+k$/i.test(cleaned)) {
    cleaned = cleaned.replace(/k$/i, '000');
  }
  
  const amount = parseFloat(cleaned);
  
  if (isNaN(amount) || amount <= 0) {
    return null;
  }
  
  // Round to 2 decimal places
  return Math.round(amount * 100) / 100;
}

/**
 * Validate phone number
 * @param {string} phone - Phone number
 * @returns {boolean} Is valid
 */
function validatePhoneNumber(phone) {
  if (!phone) return false;
  
  // Remove common formatting
  const cleaned = phone.replace(/[\s\-\(\)]/g, '');
  
  // Basic validation: should contain digits and optional + prefix
  return /^\+?\d{10,15}$/.test(cleaned);
}

/**
 * Validate category name
 * @param {string} name - Category name
 * @returns {Object} Validation result {isValid, error}
 */
function validateCategoryName(name) {
  if (!name) {
    return { isValid: false, error: 'Category name cannot be empty' };
  }
  
  const trimmed = name.trim();
  
  if (trimmed.length < 1) {
    return { isValid: false, error: 'Category name cannot be empty' };
  }
  
  if (trimmed.length > 100) {
    return { isValid: false, error: 'Category name cannot exceed 100 characters' };
  }
  
  // Check for invalid characters
  if (/[<>:"/\\|?*]/.test(trimmed)) {
    return { isValid: false, error: 'Category name contains invalid characters' };
  }
  
  return { isValid: true, error: null };
}

/**
 * Validate expense ID
 * @param {string|number} id - Expense ID
 * @returns {number|null} Parsed ID or null if invalid
 */
function validateExpenseId(id) {
  const parsed = parseInt(id);
  return isNaN(parsed) || parsed <= 0 ? null : parsed;
}

module.exports = {
  validateAmount,
  validatePhoneNumber,
  validateCategoryName,
  validateExpenseId
};
